packages/yew-router-macro/src/routable_derive.rs

```diff
@@ -22,225 +26,457 @@ impl Parse for Routable {
+    let hidden = hidden_module();
+    let mut binds = Vec::new();
+    for name in &at.names {
+        if let Some(field) = fields.iter().find(|field| &field.name == name) {
+            binds.push((field.index, BindType::String));
+        } else {
+            return Err(syn::Error::new(
+                at.span,
+                format_args!("No field named {:?}", name),
+            ));
+        }
+    }
+    let parts: TokenStream = at
+        .parts
+        .iter()
+        .map(|part| match part {
+            PathPart::Match(s) => quote! { #hidden::PathPart::Match(#s), },
+            PathPart::ExtractOne => quote! { #hidden::PathPart::ExtractOne, },
+            PathPart::ExtractAll => quote! { #hidden::PathPart::ExtractAll, },
         })
+        .collect();
+    Ok(ConstMuncher {
+        muncher: Some(quote! {
+            #hidden::PathSegmentMuncher {
+                parts: &[#parts]
+            }
+        }),
+        binds,
+    })
+}
+
+fn generate_arg_muncher(muncher: &Ident, fields: &[ParsedField]) -> syn::Result<ConstMuncher> {
+    let hidden = hidden_module();
+    let mut binds = Vec::new();
+    let mut names = Vec::new();
+    for field in fields {
+        let name = match &field.meta {
+            Some(syn::Meta::Path(_)) => field.name.clone(),
+            Some(syn::Meta::NameValue(syn::MetaNameValue {
+                lit: syn::Lit::Str(name),
+                ..
+            })) => name.value(),
+            _ => {
+                return Err(syn::Error::new(
+                    field.meta.span(),
+                    "Invalid argument binding",
+                ))
+            }
+        };
+        names.push(name);
+        binds.push((field.index, BindType::String));
     }
+    Ok(ConstMuncher {
+        muncher: Some(quote! {
+            #hidden::#muncher {
+                names: &[#(#names,)*]
+            }
+        }),
+        binds,
+    })
 }
 
-fn parse_variants_attributes(
-    variants: &Punctuated<Variant, syn::token::Comma>,
-) -> syn::Result<(Option<Ident>, Vec<LitStr>)> {
-    let mut not_founds = vec![];
-    let mut ats: Vec<LitStr> = vec![];
+fn generate_rest_glue(fields: &[ParsedField]) -> syn::Result<ConstMuncher> {
+    let binds = match fields {
+        [field] => vec![(field.index, BindType::Rest)],
+        _ => {
+            return Err(syn::Error::new(
+                Span::call_site(),
+                "Only one `rest` binding can be used",
+            ))
+        }
+    };
 
-    let mut not_found_attrs = vec![];
+    Ok(ConstMuncher {
+        muncher: None,
+        binds,
+    })
+}
 
-    for variant in variants.iter() {
-        if let Fields::Unnamed(ref field) = variant.fields {
+impl BindMode {
+    fn from_path(path: &syn::Path) -> syn::Result<Self> {
+        Ok(if path.is_ident("query_arg") {
+            Self::QueryArg
+        } else if path.is_ident("hash_arg") {
+            Self::HashArg
+        } else if path.is_ident("rest") {
+            Self::Rest
```

Comment:
Can we use `const`s instead of hard-coding these idents?

Reply:
OK, although I'm not sure it provides much benefit when this is the only place they're used.